scripts/getChangelog.py

```diff
@@ -1,90 +1,570 @@
+            7. Click OK in all windows.
+            
+            Please note that setting environment variables this way will make them available in all future terminal sessions and scripts. Use with caution.
+            """))
+          else:
+            print("Ok, moving on.")
+          break
+      except urllib.error.HTTPError as error_message:
+        if error_message.code == 401:
+          handle_exception(Exception("Invalid GitHub API token, please try again."), data)
+          token = ""
+        else:
+          handle_exception(Exception("Issue executing GraphQL query"), data)
+          token = ""
+      except Exception as error_message:
+        handle_exception(error_message, data)
+        token = ""
+    return token
+
+  if not token:
+    token = get_token()
+  # Assuming authentication is successful, we will no longer check for/catch authentication errors.
+
+  # Define base_ref
+  if not base_ref:
+    print()
+    print("Now, what should be our point of reference?")
+    print("Please enter a release code or branch")
+    print("[default: latest (in other words: previous) GitHub release]")
+    base_ref = input('Base: ')
+
+  if base_ref == "latest" or not base_ref:
+    query = '''
+    query ($owner: String!, $repo: String!) {
+      repository(owner: $owner, name: $repo) {
+        latestRelease {
+          tagName
+        }
+      }
+    }
+    '''
+    while True:
+      try:
+        variables = {"owner": owner, "repo": repo}
+        response = client.execute(query, variables)
+        data = json.loads(response)
+        base_ref = data['data']['repository']['latestRelease']['tagName']
+        print("\nOkido, let's get the latest release (" + base_ref + ")!")
+        break
+      except Exception as e:
+        handle_exception(e, data)
+        print("Does your token have enough permissions?")
+        sys.exit(1)
+  else:
+    query = '''
+    query ($owner: String!, $repo: String!, $ref: String!) {
+      repository(owner: $owner, name: $repo) {
+        ref(qualifiedName:"$ref") {
+          name
+        }
+      }
+    }
+    '''
+    while True:
+      try:
+        variables = {"owner": owner, "repo": repo, "ref": base_ref}
+        response = client.execute(query, variables)
+        data = json.loads(response)
+        if data['data']['repository']['ref'] == None:
+          print("\nError: Invalid release code or branch.")
+          print("Please try again.")
+          base_ref = input('Base: ')
+          continue
+        base_ref = data['data']['repository']['ref']['name']
+        print(f"\nNoted, {base_ref} it is.")
+        break
+      except Exception as e:
+        handle_exception(e, data)
+        print("Does your token have enough permissions?")
+        sys.exit(1)
+
+  # Define head_ref
+  if not head_ref:
+    print("\nThen, from which environment would you like to see the changes (the head)?")
```

Comment:
I don't understand `environment` here. Maybe just "where do you want the changes to start`?

Reply:
I used 'environment' because of its vagueness, and because I couldn't think of anything better. In my mental model, we're comparing a point in time (commit) or a location (branch) with another point in time or location. 'Point of reference' is a nice description of what we compare _with_, but what do we call where we are now?

In that mental model 'the changes to start' doesn't really work, because I'm comparing something in the past (the base = the previous release, represented by the tag) with something more recent (the head = what we're about to release, represented by master).

I get that 'environment' is a bit confusing in this context. Maybe this?
> Then, what would you like to get the changelog for (the head)?

Maybe with this mental model, I should also switch around asking for the head (now) and asking for the base (the past).